sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/PropertiesAccessCorrectFieldBase.cs

```diff
@@ -71,28 +73,57 @@ protected void CheckType(SymbolAnalysisContext context)
+                    if (!data.IgnoreGetter)
+                    {
+                        CheckExpectedFieldIsUsed(data.PropertySymbol.GetMethod, expectedField, data.ReadFields, context);
+                    }
+                    if (!data.IgnoreSetter)
+                    {
+                        CheckExpectedFieldIsUsed(data.PropertySymbol.SetMethod, expectedField, data.UpdatedFields, context);
+                    }
                 }
             }
         }
 
-        protected static IEnumerable<IPropertySymbol> GetExplictlyDeclaredProperties(INamedTypeSymbol symbol) =>
+        protected IEnumerable<IPropertySymbol> GetExplictlyDeclaredProperties(INamedTypeSymbol symbol) =>
             symbol.GetMembers()
                 .Where(m => m.Kind == SymbolKind.Property)
                 .OfType<IPropertySymbol>()
-                .Where(p => !p.IsImplicitlyDeclared);
+                .Where(p => ImplementsExplicitGetterOrSetter(p));
 
-        protected void CheckExpectedFieldIsUsed(IFieldSymbol expectedField, FieldData? actualField, SymbolAnalysisContext context)
+        protected void CheckExpectedFieldIsUsed(IMethodSymbol methodSymbol, IFieldSymbol expectedField, IEnumerable<FieldData> actualFields, SymbolAnalysisContext context)
         {
-            if (actualField.HasValue && actualField.Value.Field != expectedField)
+            var expectedFieldIsUsed = actualFields.Any(a => a.Field == expectedField);
+            if (!expectedFieldIsUsed || !actualFields.Any())
             {
-                context.ReportDiagnosticWhenActive(Diagnostic.Create(
-                    Rule,
-                    actualField.Value.LocationNode.GetLocation(),
-                    actualField.Value.AccessorKind == AccessorKind.Getter ? "getter" : "setter",
-                    expectedField.Name
-                    ));
+                var locationAndAccessorType = GetLocationAndAccessor(actualFields, methodSymbol);
+                if (locationAndAccessorType.Item1 != null)
+                {
+                    context.ReportDiagnosticWhenActive(Diagnostic.Create(
+                        Rule,
+                        locationAndAccessorType.Item1,
+                        locationAndAccessorType.Item2,
+                        expectedField.Name
+                        ));
+                }
+            }
+
+            Tuple<Location, string> GetLocationAndAccessor(IEnumerable<FieldData> fields, IMethodSymbol method)
+            {
+                Location location = null;
+                string accessorType = null;
+                if (fields.Count() == 1)
+                {
+                    var fieldWithValue = fields.First();
+                    location = fieldWithValue.LocationNode.GetLocation();
+                    accessorType = fieldWithValue.AccessorKind == AccessorKind.Getter ? "getter" : "setter";
+                }
+                else if (method != null)
```

Comment:
Is this code just being defensive, or are there valid cases where there are multiple fields but `method` is null?

If it's just defensive then I'd add an assertion to highlight that we've ended up in an unexpected situation.